pkg/trace/local_tracer.go

```diff
@@ -0,0 +1,203 @@
+}
+
+// NewLocalTracer creates a struct that will save all of the events passed to
+// the retuen channel to files based on their "table" (a string field in the
+// event). Each type gets its own file. The internal are purposefully not thread
+// safe to avoid the overhead of locking with each event save. Only pass events
+// to the returned channel. Call CloseAll to close all open files. Goroutine to
+// save events is started in this function.
+func NewLocalTracer(cfg *config.Config, logger log.Logger, chainID, nodeID string) (*LocalTracer, error) {
+	fm := make(map[string]*bufferedFile)
+	p := path.Join(cfg.RootDir, "data", "traces")
+	for _, table := range splitAndTrimEmpty(cfg.Instrumentation.TracingTables, ",", " ") {
+		fileName := fmt.Sprintf("%s/%s.jsonl", p, table)
+		err := os.MkdirAll(p, 0700)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create directory %s: %w", p, err)
+		}
+		file, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open or create file %s: %w", fileName, err)
+		}
+		bf := newbufferedFile(file)
+		fm[table] = bf
+	}
+
+	lt := &LocalTracer{
+		fileMap: fm,
+		cfg:     cfg,
+		canal:   make(chan Event[Entry], cfg.Instrumentation.TraceBufferSize),
+		chainID: chainID,
+		nodeID:  nodeID,
+		logger:  logger,
+	}
+
+	go lt.drainCanal()
+	if cfg.Instrumentation.TracePullAddress != "" {
+		go lt.servePullData()
+	}
+	if cfg.Instrumentation.TracePushConfig != "" {
+		s3Config, err := readS3Config(path.Join(cfg.RootDir, "config", cfg.Instrumentation.TracePushConfig))
+		if err != nil {
+			return nil, fmt.Errorf("failed to read s3 config: %w", err)
+		}
+		lt.s3Config = s3Config
+		go lt.pushLoop()
+	}
+
+	return lt, nil
+}
+
+func (lt *LocalTracer) Write(e Entry) {
+	if !lt.IsCollecting(e.Table()) {
+		return
+	}
+	lt.canal <- NewEvent(lt.chainID, lt.nodeID, e.Table(), e)
+}
+
+// ReadTable returns a file for the given table. If the table is not being
+// collected, an error is returned. This method is not thread-safe.
+func (lt *LocalTracer) ReadTable(table string) (*os.File, error) {
+	bf, has := lt.getFile(table)
+	if !has {
+		return nil, fmt.Errorf("table %s not found", table)
+	}
+
+	return bf.File()
+}
+
+func (lt *LocalTracer) IsCollecting(table string) bool {
+	if _, has := lt.getFile(table); has {
+		return true
+	}
+	return false
+}
+
+// getFile gets a file for the given type. This method is purposely
+// not thread-safe to avoid the overhead of locking with each event save.
+func (lt *LocalTracer) getFile(table string) (*bufferedFile, bool) {
+	f, has := lt.fileMap[table]
+	return f, has
+}
+
+// saveEventToFile marshals an Event into JSON and appends it to a file named after the event's Type.
+func (lt *LocalTracer) saveEventToFile(event Event[Entry]) error {
+	file, has := lt.getFile(event.Table)
+	if !has {
+		return fmt.Errorf("table %s not found", event.Table)
+	}
+
+	eventJSON, err := json.Marshal(event)
```

Comment:
Using JSON format can be inefficient because it repeats field names as keys for every event, leading to redundancy. Switching to CSV could be more efficient.

Reply:
yeah csv is definitely better, and should be a simple enough switch see

https://github.com/celestiaorg/celestia-core/pull/1279#discussion_r1550497217